### PR TITLE
Implement in-memory orchestration backend for testing

### PR DESCRIPTION
**Pull request by YunchuWang**
# Summary

## What changed?
- Added `InMemoryOrchestrationBackend` - an in-memory storage backend for orchestration state suitable for testing
- Added `TestOrchestrationClient` - a client API for scheduling and managing orchestrations against the in-memory backend
- Added `TestOrchestrationWorker` - a worker that processes orchestrations and activities using the real `OrchestrationExecutor` and `ActivityExecutor`
- Exported all testing utilities from the main package via `@microsoft/durabletask-js`
- Updated e2e tests to use the in-memory testing backend instead of requiring a gRPC sidecar
- Added comprehensive unit tests for the in-memory backend

## Why is this change needed?
- Enables unit testing and integration testing of durable orchestrations without requiring a sidecar process
- Simplifies test setup and improves test execution speed
- Allows testing orchestration logic in isolation while still exercising the real executor code paths
- Makes it easier for developers to write tests for their orchestration code

## Issues / work items
- Resolves #
- Related #

---

# Project checklist
- [ ] Release notes are not required for the next release
  - [ ] Otherwise: Notes added to `CHANGELOG.md`
- [x] Backport is not required
  - [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
- [x] All required tests have been added/updated (unit tests, E2E tests)
- [x] Breaking change?
  - [ ] If yes:
    - Impact: N/A - This is a new feature addition
    - Migration guidance: N/A

---

# AI-assisted code disclosure (required)

## Was an AI tool used? (select one)
- [ ] No
- [x] Yes, AI helped write parts of this PR (e.g., GitHub Copilot)
- [] Yes, an AI agent generated most of this PR

If AI was used:
- Tool(s): GitHub Copilot (Claude)
- AI-assisted areas/files:
  - `src/testing/in-memory-backend.ts` - Core in-memory backend implementation
  - `src/testing/test-client.ts` - Test client implementation
  - `src/testing/test-worker.ts` - Test worker implementation
  - `tests/e2e/orchestration.spec.ts` - Updated to use in-memory backend
  - `packages/durabletask-js/test/in-memory-backend.spec.ts` - Unit tests
- What you changed after AI output: Reviewed implementation, fixed continue-as-new event ordering, verified alignment with DTMB backend behavior

AI verification (required if AI was used):
- [x] I understand the code and can explain it
- [x] I verified referenced APIs/types exist and are correct
- [x] I reviewed edge cases/failure paths (timeouts, retries, cancellation, exceptions)
- [x] I reviewed concurrency/async behavior
- [x] I checked for unintended breaking or behavior changes

---

# Testing

## Automated tests
- Result: Passed (13 tests pass, 1 skipped - PurgeInstanceCriteria not supported in in-memory backend)

## Manual validation (only if runtime/behavior changed)
- Environment (OS, Node.js version, components): Windows, Node.js
- Steps + observed results:
  1. Ran `npm run test:e2e:internal` - all tests pass
  2. Verified orchestration lifecycle (create, run, complete) works correctly
  3. Verified activities, timers, sub-orchestrations, external events, termination, continue-as-new all work
- Evidence (optional): N/A

---

# Notes for reviewers
- The in-memory backend uses polling (`setImmediate`) for sub-orchestration completion notification, which differs from the production backend's event-driven approach, but achieves the same functional behavior
- `PurgeInstanceCriteria` (multi-instance purge) is not implemented - test remains skipped
- The testing implementation reuses the real `OrchestrationExecutor`, `ActivityExecutor`, and `Registry` classes, ensuring that the core orchestration logic is tested

### CODE REVIEW

File: packages/durabletask-js/src/testing/in-memory-backend.ts [hunk 1]
```diff
@@ -0,0 +1,637 @@
...
+    instance.lastUpdatedAt = new Date();
+
+    if (customStatus !== undefined) {
+      instance.customStatus = customStatus;
+    }
+
+    // Check history size limit
+    if (instance.history.length > this.maxHistorySize) {
+      throw new Error(`Orchestration '${instanceId}' exceeded maximum history size of ${this.maxHistorySize}`);
+    }
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply packages/durabletask-js/src/testing/in-memory-backend.ts]: done
